File: Solutions/0678/0678.cpp
class Solution {
public:
    bool checkValidString(string s) {
        stack <int>l;
        stack <int>x;
        int n = std::end(s) - std::begin(s);
        for (int i=0;i<n;i++){
            if (s[i] == '(') l.push(i);
            if (s[i] == '*') x.push(i);
            if (s[i] == ')'){
                if (l.size()>0) l.pop();
                else if (x.size()>0) x.pop();
                else return false;
            }
        }
        while (l.size()){
            if (x.empty()) return false;
            if (l.top() < x.top()) {
                l.pop();
                x.pop();
            }else return false;
        }
        return true;
    }
};

File: Solutions/0678/0678.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2021/9/12 23:30
# File: 0678.py
# Desc: 

class Solution:
    def checkValidString(self, s: str) -> bool:
        l = []
        x = []
        for idx,i in enumerate(s):
            if i == "(": l.append(idx)
            if i == "*": x.append(idx)
            if i == ")":
                if l: l.pop()
                elif x: x.pop()
                else: return False
        for i in l[::-1]:
            if not x or x.pop() < i: return False
        return True
